src/transaction/log_recovery_redo_parallel.cpp

```diff
@@ -517,135 +168,287 @@ namespace cublog
 
   constexpr unsigned short redo_parallel::redo_task::WAIT_AND_CHECK_MILLIS;
 
-  redo_parallel::redo_task::redo_task (std::size_t a_task_idx,
+  redo_parallel::redo_task::redo_task (std::size_t a_task_idx, bool a_do_monitor_not_applied_log_lsa,
 				       redo_parallel::task_active_state_bookkeeping &task_state_bookkeeping,
-				       redo_parallel::redo_job_queue &a_queue,
-				       const log_rv_redo_context &copy_context)
+				       const log_rv_redo_context &copy_context, std::atomic_bool &a_adding_finished)
     : m_task_idx { a_task_idx }
+    , m_do_monitor_not_applied_log_lsa { a_do_monitor_not_applied_log_lsa }
     , m_task_state_bookkeeping (task_state_bookkeeping)
-    , m_queue (a_queue)
     , m_perf_stats_definition { perf_stats_async_definition_init_list }
     , m_perf_stats { perf_stats_is_active_for_async (), m_perf_stats_definition }
     , m_redo_context { copy_context }
+    , m_produce_vec { new redo_job_vector_t () }
+  , m_adding_finished { a_adding_finished }
+  , m_not_applied_log_lsa { MAX_LSA }
   {
     // important to set this at this moment and not when execution begins
     // to circumvent race conditions where all tasks haven't yet started work
     // while already bookkeeping is being checked
     m_task_state_bookkeeping.set_active ();
+    m_in_execution.store (true);
+
+    m_produce_vec->reserve (PARALLEL_REDO_JOB_VECTOR_RESERVE_SIZE);
+  }
+
+  redo_parallel::redo_task::~redo_task ()
+  {
+    assert (is_idle ());
   }
 
   void
   redo_parallel::redo_task::execute (context_type &context)
   {
-    bool finished = false;
+    std::unique_ptr<redo_job_vector_t> jobs_vec { new redo_job_vector_t () };
+    // according to spec, reserved size survives clearing of the vector
+    // which should help to only allocate/reserve once
+    jobs_vec->reserve (PARALLEL_REDO_JOB_VECTOR_RESERVE_SIZE);
 
-    for (; !finished ;)
+    for (bool finished = false; !finished ;)
       {
-	bool adding_finished = false;
-	redo_parallel::redo_job_base *const job = m_queue.pop_job (adding_finished);
+	pop_jobs (jobs_vec);
+	const bool jobs_popped { !jobs_vec->empty ()};
 	m_perf_stats.time_and_increment (cublog::PERF_STAT_ID_PARALLEL_POP);
-
-	if (job == nullptr && adding_finished)
+	if (!jobs_popped && m_adding_finished.load ())
 	  {
 	    finished = true;
```

Comment:
In situations like these, it is usual to use `while (true)` loop and break it when it's finished.
```suggestion
	    break;
```

Reply:
Done